### ISSUE TITLE
[Quantization] After calibration I get wrong result

### ISSUE THREAD

**Issue by HeaseoChung** [NVIDIA/TensorRT#1187]
## Description

I tried to quantize ESPCN super resolution model with tensorRT. I did save my model and calibration as well without error. However, the result of output image is incorrect.

## Environment

**TensorRT Version**: 
- 7.1.3.0
**NVIDIA GPU**: 
- Jetson Xavier
**CUDA Version**: 
- 10.2
**Operating System**: 
- Ubuntu
**Python Version (if applicable)**: 
- 1.7

## Main
def main():
    parser = argparse.ArgumentParser(description="ONNX TO TENSORRT")
    parser.add_argument("--onnx", type=str, required=True, help="Insert ONNX file to convert TensorRT model.")
    args = parser.parse_args()

    onnx_file_name = args.onnx
    calibration_files = create_calibration_dataset()

    print(f'calib : {len(calibration_files)}')
    # Process 5 images at a time for calibration
    # This batch size can be different from MaxBatchSize (1 in this example)
    batchstream = calibrator.ImageBatchStream(10, calibration_files, convert_rgb_to_y)
    int8_calibrator = calibrator.PythonEntropyCalibrator(["data"], batchstream, 'temp.cache')

    with build_int8_engine(onnx_file_name, 'temp_test_db.trt', int8_calibrator) as engine, engine.create_execution_context() as context:
        evaluatate(context,image_path, 'int8')

## Model

        # Feature mapping
        self.feature_maps = nn.Sequential(
            nn.Conv2d(1, 64, kernel_size=5, stride=1, padding=2),
            nn.Tanh(),

            nn.Conv2d(64, 32, kernel_size=3, stride=1, padding=1),
            nn.Tanh()
        )

        # Sub-pixel convolution layer
        self.sub_pixel = nn.Sequential(
            nn.Conv2d(32, 1 * (scale_factor ** 2), kernel_size=3, stride=1, padding=1),
            nn.PixelShuffle(scale_factor),
            nn.Sigmoid()
        )

    def forward(self, inputs):
        out = self.feature_maps(inputs)
        out = self.sub_pixel(out)
        return out

## Building TensorRT with int8

def build_int8_engine(onnx_file_path, engine_file_path, calib):
    EXPLICIT_BATCH = 1 << (int)(trt.NetworkDefinitionCreationFlag.EXPLICIT_BATCH)

    with trt.Builder(TRT_LOGGER) as builder, builder.create_network((EXPLICIT_BATCH)) as network, trt.OnnxParser(network, TRT_LOGGER) as parser:
        # Parse model file
        print('Loading ONNX file from path {}...'.format(onnx_file_path))
        with open(onnx_file_path, 'rb') as model:
            print('Beginning ONNX file parsing')
            if not parser.parse(model.read()):
                print('ERROR: Failed to parse the ONNX file.')
                for error in range(parser.num_errors):
                    print(parser.get_error(error))
                return None
        print('Completed parsing of ONNX file')
        
        print('Network inputs:')
        for i in range(network.num_inputs):
            tensor = network.get_input(i)
            print(tensor.name, trt.nptype(tensor.dtype), tensor.shape)

        config = builder.create_builder_config()
        config.max_workspace_size = 1 << 20  # 256MiB
        if calib:
            config.set_flag(trt.BuilderFlag.INT8)
            config.int8_calibrator = calib

        print('Building an engine from file {}; this may take a while...'.format(onnx_file_path))
        engine = builder.build_engine(network, config)
        print("Completed creating Engine. Writing file to: {}".format(engine_file_path))

        with open(engine_file_path, "wb") as f:
            f.write(engine.serialize())
        return engine

## Relevant Files

[input image](https://drive.google.com/file/d/1GOSpPTeBhBaGmZqXgZ8OynVoWIGiPOle/view?usp=sharing)
[output image](https://drive.google.com/file/d/1BdDewqpPLCri2eto2QiEPlpErMuf6Qt3/view?usp=sharing)

**Comment by ttyio** [comment]
@HeaseoChung 
Some network might hard to run on INT8, you need experiment on different hyperparameter tuning,  FYI, we have a whitepaper on INT8 quantization http://arxiv.org/abs/2004.09602

Also you could check if this INT8 model works in DL framework like TF/Pytorch.

**Comment by HeaseoChung** [comment]
@ttyio 
Thank you for answering my question :)
Actually, I had done ESPCN quantization with Pytorch static quantization(PTQ) and it worked well! 
Maybe i am missing something in TensorRT but I have no way how calib in TensorRT is working.

**Comment by ttyio** [comment]
@HeaseoChung 
Is the FP32 run in TRT correct?
Is the calibration algorithm you used in Pytorch PTQ and TRT PTQ the same? 
Thanks

**Comment by HeaseoChung** [comment]
@ttyio 
I tried it with FP16 ONNX model :( 
Does it affect on image quality?

**Comment by ttyio** [comment]
@HeaseoChung 
Usually we generate INT8 engine from FP32 ONNX, else you will have 2 step floating point saturation, might loss more accuracy.

Could you generate FP32 ONNX model, and first check FP32 ONNX can return correct result in TRT, then let's calibrate using the same algorithm as PTQ in Pytorch? thanks!

**Comment by HeaseoChung** [comment]
@ttyio Thank you so much for your help!

I will do it tmr and answer back to you 👍

**Comment by HeaseoChung** [comment]
@ttyio 
Thank you so much 👍
I solved it !
